### PR TITLE
Feature/dsg 2080/count and filter

### PR DESCRIPTION
**Pull request by jcsuarezm**
# Basics

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Styling and layout changes
- [ ] Code style changes (formatting, local variables)
- [ ] Refactoring (no functional changes)
- [ ] Build or CI related changes
- [ ] Tests
- [ ] Other... Please describe:

## What changes does this PR introduce?

Changes have been implemented in both the front end and back end for the following: 

There is a new API that counts the amount of transactions that are to be shown in each dashboard tab. It optimizes to make a reduced amount of database calls. 

The API is being called by the front end to populate the tab counts. 

More complex filters are now being supported, including status, priority, and assignedToMe. Lists are also supported as values and filters can be a composition of one or more of the mentioned attributes. 

The count for the tabs is updated when changes are introduced. 

This was tested both via unit tests and manually as captured in the screenshot section

## Screenshot of the affected changes

1. Complex filter implementation used for testing: 
image

2. Call from the front end being made to the API: 
image

3. Counts displayed: 
image

## Checklist

- [ ] I have updated the documentation accordingly
- [x] I have confirmed all acceptance criteria defined by the story have been met
- [ ] I have confirmed that Sonar Qube findings for new code are reviewed and fixed, or will be whitelisted

### CODE REVIEW

File: libs/shared/data-access/work-api/src/lib/services/work-api-routes.service.ts [hunk 1]
```diff
@@ -360,4 +361,11 @@ export class WorkApiRoutesService extends HttpBaseService {
       map(dashboardsSchema => dashboardsSchema.map(dashboardSchema => new DashboardModel(dashboardSchema))),
     );
   }
+
+  /**
+   * Gets counts for each dashboard tab
+   */
+  public getDashboardCounts$(transactionSetKey: string): Observable<IDashboardCount[]> {
+    return this._handleGet$<IDashboardCount[]>(`/v1/admin/dashboards/${transactionSetKey}/counts`);
```
Comment (davidyoung77) [inline libs/shared/data-access/work-api/src/lib/services/work-api-routes.service.ts:369]: Can you return a model, the same as we are doing with other requests?

File: libs/agency/feature/dashboard/src/lib/components/dashboard/dashboard.component.ts [hunk 2]
```diff
@@ -162,10 +175,30 @@ export class DashboardComponent implements OnDestroy {
               tab.count = tab.label === currentStatus ? transaction.items.length : 0;
             });
           } else {
-            this.tabs = JSON.parse(this.tabsCopy);
+            this.tabs.forEach(tab => {
+              tab.count = this.getCountByTabLabel(tab.label);
+            });
+
+            const objectToMap = (obj: Record<string, string>): Map<string, string> => {
```
Comment (davidyoung77) [inline libs/agency/feature/dashboard/src/lib/components/dashboard/dashboard.component.ts:182]: Can you give this a more contextual name?

**Review by davidyoung77 — COMMENTED** [review]
Thanks, I left some comments also please add screenshots or video and description

File: libs/shared/data-access/work-api/src/lib/services/work-api-routes.service.ts (hunk 1 above)
Comment (jcsuarezm) [reply libs/shared/data-access/work-api/src/lib/services/work-api-routes.service.ts]: Thank you David, this has been corrected to be a model.

File: libs/agency/feature/dashboard/src/lib/components/dashboard/dashboard.component.ts (hunk 2 above)
Comment (jcsuarezm) [reply libs/agency/feature/dashboard/src/lib/components/dashboard/dashboard.component.ts]: A more descriptive name has been assigend.

File: libs/shared/data-access/work-api/src/lib/models/dashboard/dashboard-count.model.ts [hunk 3]
```diff
@@ -0,0 +1,9 @@
+export class DashboardCountModel {
```
Comment (davidyoung77) [inline libs/shared/data-access/work-api/src/lib/models/dashboard/dashboard-count.model.ts:1]: We have a patter for handling the models already

```suggestion
export class DashboardCountModel implements SchemaModel<interface> {
```

File: libs/shared/data-access/work-api/src/lib/services/work-api-routes.service.ts [hunk 4]
```diff
@@ -360,4 +361,11 @@ export class WorkApiRoutesService extends HttpBaseService {
       map(dashboardsSchema => dashboardsSchema.map(dashboardSchema => new DashboardModel(dashboardSchema))),
     );
   }
+
+  /**
+   * Gets counts for each dashboard tab
+   */
+  public getDashboardCounts$(transactionSetKey: string): Observable<DashboardCountModel[]> {
+    return this._handleGet$<DashboardCountModel[]>(`/v1/admin/dashboards/${transactionSetKey}/counts`);
```
Comment (davidyoung77) [inline libs/shared/data-access/work-api/src/lib/services/work-api-routes.service.ts:369]: You should convert to a model here
```suggestion
    return this._handleGet$<DashboardCountModel[]>(`/v1/admin/dashboards/${transactionSetKey}/counts`).pip(map => new Model);
```

File: libs/shared/data-access/work-api/src/lib/models/dashboard/dashboard-count.model.ts (hunk 3 above)
Comment (jcsuarezm) [reply libs/shared/data-access/work-api/src/lib/models/dashboard/dashboard-count.model.ts]: Thank you David, the SchemaModel interface has been implemented.

**Review by davidyoung77 — APPROVED** [review]
Thanks, looks good!

**Pull request merged**